qpdf-convert-server.py

```diff
@@ -0,0 +1,204 @@
+        send_b(data)
+
+def pdf_to_png(pagenum, pdf_path, png_path):
+    png_filename = os.path.splitext(png_path)[0]
+    cmd = ['pdftocairo', pdf_path, '-png', '-f', str(pagenum), '-l',
+           str(pagenum), '-singlefile', png_filename]
+
+    try:
+        subprocess.run(cmd, check=True)
+    except subprocess.CalledProcessError:
+        cmd = ['convert', pdf_path, f'png:{png_path}']
+        try:
+            subprocess.run(cmd, check=True)
+        except subprocess.CalledProcessError:
+            die(f'Page {pagenum} conversion failed (PDF->PNG): {err}')
+
+def png_to_rgb(pagenum, png_path, rgb_path):
+    depth = 8
+    cmd = ['convert', png_path, '-depth', str(depth), f'rgb:{rgb_path}']
+
+    try:
+        subprocess.run(cmd, check=True)
+    except subprocess.CalledProcessError:
+        die(f'Page {pagenum} conversion failed (PNG->RGB): {err}')
+
+
+###############################
+#         File-related
+###############################
+
+def create_tmp_files():
+    '''Create temporary file for storing page images and the untrusted PDF'''
+    Files = namedtuple('Files', ['pdf', 'png', 'rgb'])
+    suffixes = ('', '.png', '.rgb')
+    paths = []
+
+    for suffix in suffixes:
+        with NamedTemporaryFile(prefix='qpdf-conversion-', suffix=suffix) as f:
+            paths.append(f.name)
+
+    for path in paths:
+        with open(path, 'wb') as f:
+            f.write(b'')
+
+    return Files(pdf=paths[0], png=paths[1], rgb=paths[2])
+
+
+###############################
+#         PDF-related
+###############################
+
+def recv_pdf(pdf_path):
+    filesize = int(recvline_b().decode())
+    untrusted_data = recv_b(filesize)
+
+    with open(pdf_path, 'wb') as f:
+        f.write(untrusted_data)
+
+def get_page_count(pdf_path):
+    untrusted_pages = 0
+    output = None
+    cmd = ['pdfinfo', pdf_path]
+
+    try:
+        output = subprocess.run(cmd, capture_output=True, check=True)
+    except subprocess.CalledProcessError:
+        info(f'Probably not a PDF...')
+    else:
+        for line in output.stdout.decode().splitlines():
+            if 'Pages:' in line:
+                untrusted_pages = int(line.split(':')[1])
+
+    return untrusted_pages
+
+def process_pdf(paths):
+    page = 1
+
+    untrusted_pages = get_page_count(paths.pdf)
+    send(untrusted_pages)
+
+    while (page <= untrusted_pages):
+        pdf_to_png(page, paths.pdf, paths.png)
+        send_img_dimensions(paths.png)
+        png_to_rgb(page, paths.png, paths.rgb)
+        send_rgb_file(paths.rgb)
+        page += 1
+
+
+###############################
+#            Main
+###############################
+
+def main():
+    paths = create_tmp_files()
+
+    # FIXME:
+    #   When no more PDFs are available to process, the server will exit in
+    #   recv() (called in recv_pdf()) with an EOFError. While this works
+    #   perfectly fine, it is kinda ugly; successful runs shouldn't exit with an
+    #   error, no?
```

Comment:
This is easy to fix: catch EOFError here and break from the loop.

Reply:
I think my issue here was that handling an exception was how the server successfully exited. I didn't know if this was good practice or not or if there was another way so.

This is changed though in my local repo.